src/main/java/com/epam/aidial/core/service/ApplicationService.java

```diff
@@ -155,121 +220,266 @@ public List<Application> getApplications(ResourceDescription resource, Consumer<
-
-        if (preserveReference && application.getReference() == null) {
-            throw new IllegalArgumentException("No application reference");
-        }
-
-        if (!preserveReference && application.getReference() != null && !etag.isOverwrite()) {
-            throw new IllegalArgumentException("Creating application with provided reference is not allowed");
-        }
 
         if (application.getReference() == null) {
             application.setReference(ApplicationUtil.generateReference());
         }
 
         Application.Function function = application.getFunction();
         if (function != null) {
-            function.setStatus(null);
-            function.setTargetFolder(null);
-            function.setState(null);
+            if (application.getFeatures() == null) {
+                application.setFeatures(new Features());
+            }
+
+            application.setEndpoint(null);
+            application.getFeatures().setRateEndpoint(null);
+            application.getFeatures().setTokenizeEndpoint(null);
+            application.getFeatures().setTruncatePromptEndpoint(null);
+            application.getFeatures().setConfigurationEndpoint(null);
+            function.setAuthorBucket(resource.getBucketName());
+            function.setError(null);
+
+            if (function.getRuntime() == null) {
+                throw new IllegalArgumentException("Application function runtime must be provided");
+            }
+
+            if (function.getEnv() == null) {
+                function.setEnv(Map.of());
+            }
+
+            if (function.getMapping() == null) {
+                throw new IllegalArgumentException("Application function mapping must be provided");
+            }
+
+            verifyMapping(function.getMapping().getCompletion(), true, "Application completion mapping is missing/invalid");
+            verifyMapping(function.getMapping().getRate(), false, "Application rate mapping is invalid");
+            verifyMapping(function.getMapping().getTokenize(), false, "Application tokenize mapping is invalid");
+            verifyMapping(function.getMapping().getTruncatePrompt(), false, "Application truncate_prompt mapping is invalid");
+            verifyMapping(function.getMapping().getConfiguration(), false, "Application configuration mapping is invalid");
 
             if (function.getSourceFolder() == null) {
-                throw new IllegalArgumentException("Application function sources must be provided");
+                throw new IllegalArgumentException("Application function source folder must be provided");
             }
 
             try {
                 ResourceDescription folder = ResourceDescription.fromAnyUrl(function.getSourceFolder(), encryptionService);
 
-                if (!folder.isFolder() || folder.getType() != ResourceType.FILE) {
+                if (!folder.isFolder() || folder.getType() != ResourceType.FILE || !folder.getBucketName().equals(resource.getBucketName())) {
```

Comment:
Empty error message

Reply:
the following catch block throws a generic an error with message, so this message won't be used.